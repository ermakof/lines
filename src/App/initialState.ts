import { IAppState } from '@src/App/model/IAppState';
import createGameField from '@src/utils/createGameField';

export const GAME_FIELD_SIZE = 9;

export const INIT_GAME_LEVEL = '1';
export const INIT_GAME_FIELD_PERCENT_FILLED = 0;

interface ISettings {
  percent: number;
  chain: number;
  increase: number;
}
interface IProps {
  [key: string]: ISettings;
}
export const GAME_LEVEL_SETTINGS: IProps = {
  '1': { percent: 10, chain: 2, increase: 3 },
  '2': { percent: 20, chain: 3, increase: 4 },
  '3': { percent: 30, chain: 4, increase: 5 },
};

const initialState: IAppState = {
  userLevel: INIT_GAME_LEVEL,
  gameFieldPercentFilled: 0,
  gameFieldData: createGameField(GAME_FIELD_SIZE ** 2, INIT_GAME_FIELD_PERCENT_FILLED),
  highlightedCells: undefined,
  score: 0,
  selectedCell: undefined,
  isLoading: false,
  hitParade: undefined,
};

export default initialState;
