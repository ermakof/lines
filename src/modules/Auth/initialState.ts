import { IAuthState } from '@src/modules/Auth/model/IAuthState';

const initialState: IAuthState = {
  userProfile: undefined,
};

export default initialState;
