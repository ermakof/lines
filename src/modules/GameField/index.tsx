import React, { FC } from 'react';
import styled from '@emotion/styled';

import { ICellInfo } from '@src/modules/Cell/models';
import Cell from '@src/modules/Cell';
import { TRootState } from '@src/store';
import { useSelector } from 'react-redux';
import { GAME_FIELD_SIZE } from '@src/App/initialState';

interface IFieldContainer {
  active?: boolean;
}
export const FieldContainer = styled.section<IFieldContainer>`
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  margin: auto;
  height: fit-content;
  width: fit-content;
  opacity: ${({ active }) => (active ? 1 : 0.3)};
  min-width: 550px;
  min-height: 550px;
`;

const MessageNoData = styled.p`
  height: 80vh;
  position: relative;
  text-align: center;
`;

const GameField: FC = () => {
  const userIsLogged = useSelector(({ auth }: TRootState) => !!auth.userProfile);
  const {
    gameFieldData,
    selectedCell,
    highlightedCells = {},
  } = useSelector(({ app }: TRootState) => app);
  const widthMinus1 = GAME_FIELD_SIZE - 1;

  if (!gameFieldData.length) {
    return <MessageNoData role="noDataMessage">Нет данных!</MessageNoData>;
  }

  return (
    <FieldContainer role="cellsGrid" active={userIsLogged}>
      {gameFieldData.map((item: ICellInfo, index: number) => {
        const y = Math.floor(index / GAME_FIELD_SIZE);
        const x = index % GAME_FIELD_SIZE;
        const isRight = x === widthMinus1;
        const isLeft = x === 0;
        return (
          <Cell
            key={`${y}-${x}`}
            num={index}
            filled={item}
            isSelected={selectedCell === index}
            isLeft={isLeft}
            isRight={isRight}
            highlighted={highlightedCells[index]}
          />
        );
      })}
      <div style={{ clear: 'both' }} />
    </FieldContainer>
  );
};

export default GameField;
