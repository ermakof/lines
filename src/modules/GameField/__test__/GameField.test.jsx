import React from 'react';
import { Provider } from 'react-redux';
import { render, screen } from '@testing-library/react';
import configureMockStore from 'redux-mock-store';
import { useSelector, useDispatch } from 'react-redux';

import GameField from '../index';
import initialState from "@src/store/initialState";

const mockStore = configureMockStore();
let store;

jest.mock('react-redux', () => ({
  __esModule: true,
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

describe('GameField', () => {
  beforeEach(() => {
    store = mockStore(initialState);
  });

  it('No render cells with gameFieldSize = 0', () => {
    const mockDispatch = jest.fn();
    useDispatch.mockReturnValue(mockDispatch);
    useSelector.mockReturnValue({
      gameFieldData: [],
      gameFieldSize: 0,
    });
    const { asFragment } = render(
      <Provider store={store}>
        <GameField />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
    const message = screen.getByRole(/^noDataMessage$/gi);
    expect(message).toBeInTheDocument();
  });

  it('renders 9 cells from data:[9 items]', () => {
    const mockDispatch = jest.fn();
    useDispatch.mockReturnValue(mockDispatch);
    useSelector.mockReturnValue({
      gameFieldSize: 3,
      gameFieldPercentFilled: 10,
      gameFieldData: [1, 0, 0, 0, 0, 0, 0, 0, 0],
    });
    const { asFragment } = render(
      <Provider store={store}>
        <GameField />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
    const cellsGrid = screen.getByRole(/cellsGrid/gi);
    expect(cellsGrid).toBeInTheDocument();
    const cellContainer = screen.getAllByRole(/cellContainer/gi);
    expect(cellContainer.length).toBe(5);
  });
});
