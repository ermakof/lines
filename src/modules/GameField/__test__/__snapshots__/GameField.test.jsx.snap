// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameField No render cells, game field size = 0 1`] = `
<DocumentFragment>
  <p
    class="css-5evfw1"
    role="noDataMessage"
  >
    Нет данных!
  </p>
</DocumentFragment>
`;

exports[`GameField renders 9 cells from data:[9 items] 1`] = `
<DocumentFragment>
  <section
    class="css-1werspr"
    role="cellsGrid"
  >
    <div
      class="css-es8rv7"
      role="cellContainer-0"
    >
      <div
        class="css-zzdbx3"
        role="cellContent-0"
      />
    </div>
    <div
      class="css-2r18jt"
      role="cellContainer-1"
    />
    <div
      class="css-2r18jt"
      role="cellContainer-2"
    />
    <div
      class="css-2r18jt"
      role="cellContainer-3"
    />
    <div
      class="css-2r18jt"
      role="cellContainer-4"
    />
    <div
      class="css-2r18jt"
      role="cellContainer-5"
    />
    <div
      class="css-2r18jt"
      role="cellContainer-6"
    />
    <div
      class="css-2r18jt"
      role="cellContainer-7"
    />
    <div
      class="css-1jbr29x"
      role="cellContainer-8"
    />
    <div
      style="clear: both;"
    />
  </section>
</DocumentFragment>
`;
